pkg/placement/raft/server.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// Server is Raft server implementation.
+type Server struct {
+	id  string
+	fsm *FSM
+
+	bootstrap bool
+	inMem     bool
+	raftBind  string
+	peers     []PeerInfo
+
+	raft          *raft.Raft
+	raftStore     *raftboltdb.BoltStore
+	raftTransport *raft.NetworkTransport
+	raftInmem     *raft.InmemStore
+}
+
+// New creates Raft server node.
+func New(id string, inMem, bootstrap bool, peers []PeerInfo) *Server {
+	raftBind := raftAddressForID(id, peers)
+	if raftBind == "" {
+		return nil
+	}
+
+	return &Server{
+		id:        id,
+		inMem:     inMem,
+		bootstrap: bootstrap,
+		raftBind:  raftBind,
+		peers:     peers,
+	}
+}
+
+// StartRaft starts Raft node with Raft protocol configuration. if config is nil,
+// the default config will be used.
+func (s *Server) StartRaft(config *raft.Config) error {
+	// If we have an unclean exit then attempt to close the Raft store.
+	defer func() {
+		if s.raft == nil && s.raftStore != nil {
+			if err := s.raftStore.Close(); err != nil {
+				logging.Errorf("failed to close log storage: %v", err)
+			}
+		}
+	}()
+
+	s.fsm = newFSM()
+
+	// TODO: replace tls enabled transport layer using workload cert
+	addr, err := net.ResolveTCPAddr("tcp", s.raftBind)
+	if err != nil {
+		return err
+	}
+
+	trans, err := raft.NewTCPTransport(s.raftBind, addr, 3, 10*time.Second, os.Stderr)
+	if err != nil {
+		return err
+	}
+
+	s.raftTransport = trans
+
+	// Build an all in-memory setup for dev mode, otherwise prepare a full
+	// disk-based setup.
+	var logStore raft.LogStore
+	var stable raft.StableStore
+	var snap raft.SnapshotStore
+
+	if s.inMem {
+		s.raftInmem = raft.NewInmemStore()
+		stable = s.raftInmem
+		logStore = s.raftInmem
+		snap = raft.NewInmemSnapshotStore()
+	} else {
+		if err = ensureDir(s.raftStorePath()); err != nil {
+			return errors.Wrap(err, "failed to create log store directory")
+		}
+
+		// Create the backend raft store for logs and stable storage.
+		s.raftStore, err = raftboltdb.NewBoltStore(filepath.Join(s.raftStorePath(), "raft.db"))
+		if err != nil {
+			return err
+		}
+		stable = s.raftStore
+
+		// Wrap the store in a LogCache to improve performance.
+		logStore, err = raft.NewLogCache(raftLogCacheSize, s.raftStore)
+		if err != nil {
+			return err
+		}
+
+		// Create the snapshot store.
+		snap, err = raft.NewFileSnapshotStore(s.raftStorePath(), snapshotsRetained, os.Stderr)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Setup Raft configuration.
+	if config == nil {
+		// Set default configuration for raft
+		config = &raft.Config{
+			ProtocolVersion:    raft.ProtocolVersionMax,
+			HeartbeatTimeout:   1000 * time.Millisecond,
+			ElectionTimeout:    1000 * time.Millisecond,
+			CommitTimeout:      50 * time.Millisecond,
+			MaxAppendEntries:   64,
+			ShutdownOnRemove:   true,
+			TrailingLogs:       10240,
+			SnapshotInterval:   120 * time.Second,
+			SnapshotThreshold:  8192,
+			LeaderLeaseTimeout: 500 * time.Millisecond,
+		}
+	}
+
+	config.LogLevel = "DEBUG"
```

Comment:
This is too verbose. INFO should be OK.

Reply:
I will add comment later. but, this option is for the internal logger,
but if you look at the code, this PR create adaptor logger to use dapr logger. otherwise, the internal logger's log format is inconsistent with dapr. 
so I set lowest internal log level to let dapr logger decide whether logging or not.

Reply:
I added the comments.